File: moe/natj/natj/natj-objctests/src/test/java/objc/binding/ObjCBlocks.java
/*
Copyright 2014-2016 Intel Corporation

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package objc.binding;


import org.moe.natj.general.NatJ;
import org.moe.natj.general.Pointer;
import org.moe.natj.general.ann.*;
import org.moe.natj.general.ann.Runtime;
import org.moe.natj.general.ptr.*;
import org.moe.natj.objc.ObjCRuntime;
import org.moe.natj.objc.ann.ObjCBlock;
import org.moe.natj.objc.ann.ObjCClassBinding;
import org.moe.natj.objc.ann.Selector;
import apple.NSObject;
import apple.coregraphics.struct.CGPoint;
import apple.coregraphics.struct.CGSize;
import apple.foundation.NSError;
import apple.foundation.NSMutableArray;

@Generated
@Runtime(ObjCRuntime.class)
@ObjCClassBinding
@Library("TestClassesObjC")
public class ObjCBlocks extends NSObject {
    static {
        NatJ.register();
    }

    @Generated
    protected ObjCBlocks(Pointer peer) {
        super(peer);
    }

    @Generated
    @Owned
    @Selector("alloc")
    public static native ObjCBlocks alloc();

    @Generated
    @Selector("execBlockWithBool:")
    public static native void execBlockWithBool(
            @ObjCBlock(name = "call_execBlockWithBool") Block_execBlockWithBool value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithBool {
        @Generated
        void call_execBlockWithBool(boolean arg0);
    }

    @Generated
    @Selector("execBlockWithBoolPtr:")
    public static native void execBlockWithBoolPtr(
            @ObjCBlock(name = "call_execBlockWithBoolPtr") Block_execBlockWithBoolPtr value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithBoolPtr {
        @Generated
        void call_execBlockWithBoolPtr(BoolPtr arg0);
    }

    @Generated
    @Selector("execBlockWithByte:")
    public static native void execBlockWithByte(
            @ObjCBlock(name = "call_execBlockWithByte") Block_execBlockWithByte value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithByte {
        @Generated
        void call_execBlockWithByte(byte arg0);
    }

    @Generated
    @Selector("execBlockWithBytePtr:")
    public static native void execBlockWithBytePtr(
            @ObjCBlock(name = "call_execBlockWithBytePtr") Block_execBlockWithBytePtr value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithBytePtr {
        @Generated
        void call_execBlockWithBytePtr(BytePtr arg0);
    }

    @Generated
    @Selector("execBlockWithChar:")
    public static native void execBlockWithChar(
            @ObjCBlock(name = "call_execBlockWithChar") Block_execBlockWithChar value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithChar {
        @Generated
        void call_execBlockWithChar(char arg0);
    }

    @Generated
    @Selector("execBlockWithCharPtr:")
    public static native void execBlockWithCharPtr(
            @ObjCBlock(name = "call_execBlockWithCharPtr") Block_execBlockWithCharPtr value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithCharPtr {
        @Generated
        void call_execBlockWithCharPtr(CharPtr arg0);
    }

    @Generated
    @Selector("execBlockWithConstBoolPtr:")
    public static native void execBlockWithConstBoolPtr(
            @ObjCBlock(name = "call_execBlockWithConstBoolPtr") Block_execBlockWithConstBoolPtr value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithConstBoolPtr {
        @Generated
        void call_execBlockWithConstBoolPtr(ConstBoolPtr arg0);
    }

    @Generated
    @Selector("execBlockWithConstBytePtr:")
    public static native void execBlockWithConstBytePtr(
            @ObjCBlock(name = "call_execBlockWithConstBytePtr") Block_execBlockWithConstBytePtr value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithConstBytePtr {
        @Generated
        void call_execBlockWithConstBytePtr(ConstBytePtr arg0);
    }

    @Generated
    @Selector("execBlockWithConstCharPtr:")
    public static native void execBlockWithConstCharPtr(
            @ObjCBlock(name = "call_execBlockWithConstCharPtr") Block_execBlockWithConstCharPtr value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithConstCharPtr {
        @Generated
        void call_execBlockWithConstCharPtr(ConstCharPtr arg0);
    }

    @Generated
    @Selector("execBlockWithConstDoublePtr:")
    public static native void execBlockWithConstDoublePtr(
            @ObjCBlock(name = "call_execBlockWithConstDoublePtr") Block_execBlockWithConstDoublePtr value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithConstDoublePtr {
        @Generated
        void call_execBlockWithConstDoublePtr(ConstDoublePtr arg0);
    }

    @Generated
    @Selector("execBlockWithConstFloatPtr:")
    public static native void execBlockWithConstFloatPtr(
            @ObjCBlock(name = "call_execBlockWithConstFloatPtr") Block_execBlockWithConstFloatPtr value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithConstFloatPtr {
        @Generated
        void call_execBlockWithConstFloatPtr(ConstFloatPtr arg0);
    }

    @Generated
    @Selector("execBlockWithConstIntPtr:")
    public static native void execBlockWithConstIntPtr(
            @ObjCBlock(name = "call_execBlockWithConstIntPtr") Block_execBlockWithConstIntPtr value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithConstIntPtr {
        @Generated
        void call_execBlockWithConstIntPtr(ConstIntPtr arg0);
    }

    @Generated
    @Selector("execBlockWithConstLongPtr:")
    public static native void execBlockWithConstLongPtr(
            @ObjCBlock(name = "call_execBlockWithConstLongPtr") Block_execBlockWithConstLongPtr value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithConstLongPtr {
        @Generated
        void call_execBlockWithConstLongPtr(ConstLongPtr arg0);
    }

    @Generated
    @Selector("execBlockWithConstShortPtr:")
    public static native void execBlockWithConstShortPtr(
            @ObjCBlock(name = "call_execBlockWithConstShortPtr") Block_execBlockWithConstShortPtr value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithConstShortPtr {
        @Generated
        void call_execBlockWithConstShortPtr(ConstShortPtr arg0);
    }

    @Generated
    @Selector("execBlockWithConstStructPtr:")
    public static native void execBlockWithConstStructPtr(
            @ObjCBlock(name = "call_execBlockWithConstStructPtr") Block_execBlockWithConstStructPtr value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithConstStructPtr {
        @Generated
        void call_execBlockWithConstStructPtr(
                @ReferenceInfo(type = CGSize.class) ConstPtr<CGSize> arg0);
    }

    @Generated
    @Selector("execBlockWithDouble:")
    public static native void execBlockWithDouble(
            @ObjCBlock(name = "call_execBlockWithDouble") Block_execBlockWithDouble value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithDouble {
        @Generated
        void call_execBlockWithDouble(double arg0);
    }

    @Generated
    @Selector("execBlockWithDoublePtr:")
    public static native void execBlockWithDoublePtr(
            @ObjCBlock(name = "call_execBlockWithDoublePtr") Block_execBlockWithDoublePtr value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithDoublePtr {
        @Generated
        void call_execBlockWithDoublePtr(DoublePtr arg0);
    }

    @Generated
    @Selector("execBlockWithFloat:")
    public static native void execBlockWithFloat(
            @ObjCBlock(name = "call_execBlockWithFloat") Block_execBlockWithFloat value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithFloat {
        @Generated
        void call_execBlockWithFloat(float arg0);
    }

    @Generated
    @Selector("execBlockWithFloatPtr:")
    public static native void execBlockWithFloatPtr(
            @ObjCBlock(name = "call_execBlockWithFloatPtr") Block_execBlockWithFloatPtr value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithFloatPtr {
        @Generated
        void call_execBlockWithFloatPtr(FloatPtr arg0);
    }

    @Generated
    @Selector("execBlockWithInt:")
    public static native void execBlockWithInt(
            @ObjCBlock(name = "call_execBlockWithInt") Block_execBlockWithInt value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithInt {
        @Generated
        void call_execBlockWithInt(int arg0);
    }

    @Generated
    @Selector("execBlockWithIntPtr:")
    public static native void execBlockWithIntPtr(
            @ObjCBlock(name = "call_execBlockWithIntPtr") Block_execBlockWithIntPtr value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithIntPtr {
        @Generated
        void call_execBlockWithIntPtr(IntPtr arg0);
    }

    @Generated
    @Selector("execBlockWithLong:")
    public static native void execBlockWithLong(
            @ObjCBlock(name = "call_execBlockWithLong") Block_execBlockWithLong value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithLong {
        @Generated
        void call_execBlockWithLong(long arg0);
    }

    @Generated
    @Selector("execBlockWithLongPtr:")
    public static native void execBlockWithLongPtr(
            @ObjCBlock(name = "call_execBlockWithLongPtr") Block_execBlockWithLongPtr value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithLongPtr {
        @Generated
        void call_execBlockWithLongPtr(LongPtr arg0);
    }

    @Generated
    @Selector("execBlockWithShort:")
    public static native void execBlockWithShort(
            @ObjCBlock(name = "call_execBlockWithShort") Block_execBlockWithShort value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithShort {
        @Generated
        void call_execBlockWithShort(short arg0);
    }

    @Generated
    @Selector("execBlockWithShortPtr:")
    public static native void execBlockWithShortPtr(
            @ObjCBlock(name = "call_execBlockWithShortPtr") Block_execBlockWithShortPtr value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithShortPtr {
        @Generated
        void call_execBlockWithShortPtr(ShortPtr arg0);
    }

    @Generated
    @Selector("execBlockWithStruct:")
    public static native void execBlockWithStruct(
            @ObjCBlock(name = "call_execBlockWithStruct") Block_execBlockWithStruct value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithStruct {
        @Generated
        void call_execBlockWithStruct(@ByValue CGSize arg0);
    }

    @Generated
    @Selector("execBlockWithStructPtr2:")
    public static native void execBlockWithStructPtr2(
            @ObjCBlock(name = "call_execBlockWithStructPtr2") Block_execBlockWithStructPtr2 value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithStructPtr2 {
        @Generated
        void call_execBlockWithStructPtr2(long arg0, CGPoint arg1);
    }

    @Generated
    @Selector("execBlockWithStructPtr:")
    public static native void execBlockWithStructPtr(
            @ObjCBlock(name = "call_execBlockWithStructPtr") Block_execBlockWithStructPtr value);

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_execBlockWithStructPtr {
        @Generated
        void call_execBlockWithStructPtr(
                @ReferenceInfo(type = CGSize.class) Ptr<CGSize> arg0);
    }

    @Generated
    @Selector("getBlockWithOutArg")
    @ObjCBlock(name = "call_getBlockWithOutArg_ret")
    public static native Block_getBlockWithOutArg_ret getBlockWithOutArg();

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_getBlockWithOutArg_ret {
        @Generated
        void call_getBlockWithOutArg_ret(
                @ReferenceInfo(type = NSError.class) Ptr<NSError> arg0);
    }

    @Generated
    @Selector("getMultipleBlock")
    @ObjCBlock(name = "call_getMultipleBlock_ret")
    public static native Block_getMultipleBlock_ret getMultipleBlock();

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_getMultipleBlock_ret {
        @Generated
        boolean call_getMultipleBlock_ret(boolean bo1, byte by1, char ch1, double do1, float fl1, int in1, long lo1, short sh1, @ByValue CGSize si1,
                                          boolean bo2, byte by2, char ch2, double do2, float fl2, int in2, long lo2, short sh2, @ByValue CGSize si2);
    }

    @Generated
    @Selector("getBooleanAnderBlock")
    @ObjCBlock(name = "call_getBooleanAnderBlock_ret")
    public static native Block_getBooleanAnderBlock_ret getBooleanAnderBlock();

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_getBooleanAnderBlock_ret {
        @Generated
        boolean call_getBooleanAnderBlock_ret(boolean arg0, boolean arg1);
    }

    @Generated
    @Selector("getByteMultiplierBlock")
    @ObjCBlock(name = "call_getByteMultiplierBlock_ret")
    public static native Block_getByteMultiplierBlock_ret getByteMultiplierBlock();

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_getByteMultiplierBlock_ret {
        @Generated
        byte call_getByteMultiplierBlock_ret(byte arg0, byte arg1);
    }

    @Generated
    @Selector("getCharMultiplierBlock")
    @ObjCBlock(name = "call_getCharMultiplierBlock_ret")
    public static native Block_getCharMultiplierBlock_ret getCharMultiplierBlock();

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_getCharMultiplierBlock_ret {
        @Generated
        char call_getCharMultiplierBlock_ret(char arg0, char arg1);
    }

    @Generated
    @Selector("getDoubleMultiplierBlock")
    @ObjCBlock(name = "call_getDoubleMultiplierBlock_ret")
    public static native Block_getDoubleMultiplierBlock_ret getDoubleMultiplierBlock();

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_getDoubleMultiplierBlock_ret {
        @Generated
        double call_getDoubleMultiplierBlock_ret(double arg0, double arg1);
    }

    @Generated
    @Selector("getFloatMultiplierBlock")
    @ObjCBlock(name = "call_getFloatMultiplierBlock_ret")
    public static native Block_getFloatMultiplierBlock_ret getFloatMultiplierBlock();

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_getFloatMultiplierBlock_ret {
        @Generated
        float call_getFloatMultiplierBlock_ret(float arg0, float arg1);
    }

    @Generated
    @Selector("getIntMultiplierBlock")
    @ObjCBlock(name = "call_getIntMultiplierBlock_ret")
    public static native Block_getIntMultiplierBlock_ret getIntMultiplierBlock();

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_getIntMultiplierBlock_ret {
        @Generated
        int call_getIntMultiplierBlock_ret(int arg0, int arg1);
    }

    @Generated
    @Selector("getLongMultiplierBlock")
    @ObjCBlock(name = "call_getLongMultiplierBlock_ret")
    public static native Block_getLongMultiplierBlock_ret getLongMultiplierBlock();

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_getLongMultiplierBlock_ret {
        @Generated
        long call_getLongMultiplierBlock_ret(long arg0, long arg1);
    }

    @Generated
    @Selector("getNFloatAdder")
    @ObjCBlock(name = "call_getNFloatAdder_ret")
    public static native Block_getNFloatAdder_ret getNFloatAdder();

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_getNFloatAdder_ret {
        @Generated
        @NFloat
        double call_getNFloatAdder_ret(@NFloat double arg0, @NFloat double arg1);
    }

    @Generated
    @Selector("getNIntAdder")
    @ObjCBlock(name = "call_getNIntAdder_ret")
    public static native Block_getNIntAdder_ret getNIntAdder();

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_getNIntAdder_ret {
        @Generated
        @NInt
        long call_getNIntAdder_ret(@NInt long arg0, @NInt long arg1);
    }

    @Generated
    @Selector("getNUIntAdder")
    @ObjCBlock(name = "call_getNUIntAdder_ret")
    public static native Block_getNUIntAdder_ret getNUIntAdder();

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_getNUIntAdder_ret {
        @Generated
        @NUInt
        long call_getNUIntAdder_ret(@NUInt long arg0, @NUInt long arg1);
    }

    @Generated
    @Selector("getShortMultiplierBlock")
    @ObjCBlock(name = "call_getShortMultiplierBlock_ret")
    public static native Block_getShortMultiplierBlock_ret getShortMultiplierBlock();

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_getShortMultiplierBlock_ret {
        @Generated
        short call_getShortMultiplierBlock_ret(short arg0, short arg1);
    }

    @Generated
    @Selector("getStringArrayElementAdderBlock")
    @ObjCBlock(name = "call_getStringArrayElementAdderBlock_ret")
    public static native Block_getStringArrayElementAdderBlock_ret getStringArrayElementAdderBlock();

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_getStringArrayElementAdderBlock_ret {
        @Generated
        NSMutableArray<?> call_getStringArrayElementAdderBlock_ret(String arg0,
                                                                   String arg1);
    }

    @Generated
    @Selector("getStringConcatenaterBlock")
    @ObjCBlock(name = "call_getStringConcatenaterBlock_ret")
    public static native Block_getStringConcatenaterBlock_ret getStringConcatenaterBlock();

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_getStringConcatenaterBlock_ret {
        @Generated
        String call_getStringConcatenaterBlock_ret(String arg0, String arg1);
    }

    @Generated
    @Selector("getVoidStringArrayElementAdderBlock")
    @ObjCBlock(name = "call_getVoidStringArrayElementAdderBlock_ret")
    public static native Block_getVoidStringArrayElementAdderBlock_ret getVoidStringArrayElementAdderBlock();

    @Runtime(ObjCRuntime.class)
    @Generated
    public interface Block_getVoidStringArrayElementAdderBlock_ret {
        @Generated
        void call_getVoidStringArrayElementAdderBlock_ret(
                NSMutableArray<?> arg0, String arg1, String arg2);
    }

    @Generated
    @Selector("init")
    public native ObjCBlocks init();

    @Generated
    @Selector("validateExecBlockWithStructPtr2:")
    public static native boolean validateExecBlockWithStructPtr2(
            @UncertainArgument("Options: reference, array Fallback: reference") CGPoint value);
}

File: natj/natj/natj-objctests/src/test/native/ObjCBlocks.m
/*
Copyright 2014-2016 Intel Corporation

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#import "ObjCBlocks.h"

typedef struct {
    CGPoint pt;
    uint64_t extra;
} CGPointExtended;

@implementation ObjCBlocks

+ (bool (^)(bool, int8_t, uint16_t, double, float, int32_t, int64_t, int16_t, CGSize,
            bool, int8_t, uint16_t, double, float, int32_t, int64_t, int16_t, CGSize))getMultipleBlock {
    return ^bool (bool bo1, int8_t by1, uint16_t ch1, double do1, float fl1, int32_t in1, int64_t lo1, int16_t sh1, CGSize si1,
                  bool bo2, int8_t by2, uint16_t ch2, double do2, float fl2, int32_t in2, int64_t lo2, int16_t sh2, CGSize si2) {
        return (bo1 == bo2) && (by1 == by2) && (ch1 == ch2) && (do1 == do2) && (fl1 == fl2) && (in1 == in2) && (lo1 == lo2) && (sh1 == sh2) && (bo1 == bo2) && (si1.width == si2.width) && (si1.height == si2.height);
    };
}

+ (bool (^)(bool, bool))getBooleanAnderBlock {
    return ^bool (bool a, bool b) {
        return a && b;
    };
}

+ (int8_t (^)(int8_t, int8_t))getByteMultiplierBlock {
    return ^int8_t(int8_t a, int8_t b) {
        return a * b;
    };
}

+ (unichar (^)(unichar, unichar))getCharMultiplierBlock {
    return ^unichar (unichar a, unichar b) {
        return a * b;
    };
}

+ (int16_t (^)(int16_t, int16_t))getShortMultiplierBlock {
    return ^int16_t (int16_t a, int16_t b) {
        return a * b;
    };
}

+ (int32_t (^)(int32_t, int32_t))getIntMultiplierBlock {
    return ^int32_t (int32_t a, int32_t b) {
        return a * b;
    };
}

+ (int64_t (^)(int64_t, int64_t))getLongMultiplierBlock {
    return ^int64_t (int64_t a, int64_t b) {
        return a * b;
    };
}

+ (float (^)(float, float))getFloatMultiplierBlock {
    return ^float (float a, float b) {
        return a * b;
    };
}

+ (double (^)(double, double))getDoubleMultiplierBlock {
    return ^double (double a, double b) {
        return a * b;
    };
}

+ (void (^)(NSMutableArray*, NSString*, NSString*))getVoidStringArrayElementAdderBlock {
    return ^void (NSMutableArray* array, NSString* a, NSString* b) {
        [array addObject:a];
        [array addObject:b];
    };
}

+ (NSMutableArray* (^)(NSString*, NSString*))getStringArrayElementAdderBlock {
    return ^NSMutableArray* (NSString* a, NSString* b) {
        return [NSMutableArray arrayWithObjects:a, b, nil];
    };
}

+ (NSString* (^)(NSString*, NSString*))getStringConcatenaterBlock {
    return ^NSString* (NSString* a, NSString* b) {
        return [a stringByAppendingString:b];
    };
}

+ (NSInteger (^)(NSInteger, NSInteger))getNIntAdder {
    return ^NSInteger(NSInteger a, NSInteger b) {
        return a + b;
    };
}

+ (NSUInteger (^)(NSUInteger, NSUInteger))getNUIntAdder {
    return ^NSUInteger(NSUInteger a, NSUInteger b) {
        return a + b;
    };
}

+ (CGFloat (^)(CGFloat, CGFloat))getNFloatAdder {
    return ^CGFloat(CGFloat a, CGFloat b) {
        return a + b;
    };
}

+ (void (^)(NSError**))getBlockWithOutArg {
    return ^(NSError** error) {
        error[0] = [[[NSError alloc] initWithDomain:@"test" code:123 userInfo:nil] autorelease];
    };
}

#define MOE_NATJGEN_TEST_BLOCK_EXEC_DEF(__name, __type, __value)\
+ (void)execBlockWith ## __name:(void (^)(__type))value { value((__type)__value); }

#define MOE_NATJGEN_TEST_BLOCK_EXEC_DEF2(__name, __type)\
+ (void)execBlockWith ## __name:(void (^)(__type*))value {\
__type* var = calloc(sizeof(__type), 1);\
value(var);\
free(var);\
}

#define MOE_NATJGEN_TEST_BLOCK_EXEC_DEF3(__name, __type)\
+ (void)execBlockWith ## __name:(void (^)(const __type*))value {\
const __type* var = calloc(sizeof(__type), 1);\
value(var);\
free((void*)var);\
}

MOE_NATJGEN_TEST_BLOCK_EXEC_DEF(Bool, bool, 1)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF(Byte, int8_t, 0xFA)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF(Short, int16_t, 0xFAFA)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF(Char, unichar, 0xFAFA)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF(Int, int32_t, 0xFAFAFAFA)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF(Long, int64_t, 0xFAFAFAFAFAFAFAFA)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF(Float, float, 3.14f)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF(Double, double, 3.14)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF(Struct, CGSize, CGSizeMake(10.0, 20.0))

MOE_NATJGEN_TEST_BLOCK_EXEC_DEF2(BoolPtr, bool)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF2(BytePtr, int8_t)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF2(ShortPtr, int16_t)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF2(CharPtr, unichar)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF2(IntPtr, int32_t)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF2(LongPtr, int64_t)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF2(FloatPtr, float)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF2(DoublePtr, double)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF2(StructPtr, CGSize)

+ (void)execBlockWithStructPtr2:(void (^)(uint64_t, CGPoint*))value {
    CGPointExtended* var = calloc(sizeof(CGPointExtended), 1);
    var->pt = CGPointMake(10.0, 20.0);
    var->extra = 0xFEFEFEFEFEFEFEFEL;
    value((uint64_t)var, (CGPoint*)var);
    free(var);
}

+ (bool)validateExecBlockWithStructPtr2:(CGPoint*)value {
    return ((CGPointExtended*)value)->extra == 0xFEFEFEFEFEFEFEFEL;
}

MOE_NATJGEN_TEST_BLOCK_EXEC_DEF3(ConstBoolPtr, bool)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF3(ConstBytePtr, int8_t)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF3(ConstShortPtr, int16_t)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF3(ConstCharPtr, unichar)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF3(ConstIntPtr, int32_t)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF3(ConstLongPtr, int64_t)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF3(ConstFloatPtr, float)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF3(ConstDoublePtr, double)
MOE_NATJGEN_TEST_BLOCK_EXEC_DEF3(ConstStructPtr, CGSize)

@end
